### PR TITLE
Feature/74 branding

### LINKED ISSUE #74: Create a form for choosing basic brand elements, for example, colors and logo
For the purpose of re-branding / whitelabeling Apinf, there should be an option for modifying colors and logos. This may be related to #327.
User story: As Harri (API administrator), I want to add brand elements to our organisations Apinf installation

# Task

## Choose
* [x] Choose what UI elements to make editable through configuration page
* [x] Determine how to make UI elements editable
  * [x] E.g. template variables for CSS colors
  * [x] E.g. template variable for logo

# Design
**Do not skip this step**
* [x] Create design document describing the intended configuration page
* [x] Create wireframe for intended configuration section design
* [x] Include wireframe in design document and this issue
* [x] Publish wireframe(s) in design documentation section of project sourcecode

# Implement
* [x] Add a branding section / tab / page to the configuration page created in #327
* [x] Create a schema for branding elements
  * [x] Colors
  * [x] Logo
* [x] Create necessary collection(s)
  * [x] FS collection for branding image
  * [x] Field(s) in Configuration collection for branding

### PR DESCRIPTION
**Pull request by elnzv**
Closes #74

### CODE REVIEW

File: client/layouts/master_layout/navbar/navbar.js [hunk 1]
```diff
@@ -4,5 +4,12 @@ Template.navbar.helpers({
     var profilePicture = ProfilePictures.findOne({});
     // return that url
     return profilePicture.url();
+  },
+  projectLogo: function () {
+    var lastUploadedLogo = ProjectLogo.find({}, {sort: {uploadedAt: -1}}).fetch()[0];
```
Comment (frenchbread) [inline client/layouts/master_layout/navbar/navbar.js]: @elnzv Would you also filter by user's id who has uploaded the logo? Otherwise it looks like a bug (or a feature :)).

File: client/views/branding/branding.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    // Get Branding collection content
+    return Branding.findOne();
+  }
+});
+
+
+Template.branding.helpers({
+  projectLogo: function () {
+    // Get last uploaded image from collection
+    var lastUploadedLogo = ProjectLogo.find({}, {sort: {uploadedAt: -1}}).fetch()[0];
```
Comment (frenchbread) [inline client/views/branding/branding.js]: Same here.

File: lib/router/jsonRoutes.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
+JsonRoutes.add("get", "api/:id/swagger.json", function (request, response, next) {
```
Comment (frenchbread) [inline lib/router/jsonRoutes.js]: Why is this file here? It should not be related to this PR.

**Comment by frenchbread** [comment]
Also as noted in a parent issue:

1. Is it possible to make a logo be located in the middle vertically? (Now it's aligned by the top edge) Otherwise provide a user with dimensions needed for logo image.

2. File extension is not checked, so I was able to upload `.zip` file.

3. Color schema is not applied when the selection in the select-box changes.

File: client/layouts/master_layout/navbar/navbar.js (hunk 1 above)
Comment (brylie) [inline client/layouts/master_layout/navbar/navbar.js]: There should be only one `last uploaded logo` file. The user may not be relevant, as long as they had permission to change the logo. Illya, make sure only admin users can modify these settings please.

File: client/views/branding/branding.js (hunk 2 above)
Comment (brylie) [inline client/views/branding/branding.js]: Only allowing admin users to upload project logo may make filtering by user unnecessary.

**Comment by frenchbread** [comment]
I've added suggested changes.

@apinf/developers Please review & merge this.

File: both/collections/branding.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    ],
+    label: 'Choose you color theme:'
+  }
+});
+
+Branding.attachSchema(Schemas.BrandingSchema);
+
+Branding.allow({
+  insert: function () {
+    return true;
```
Comment (brylie) [inline both/collections/branding.js]: Add check here to make sure user has Role === 'admin'

File: both/collections/branding.js [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+  },
+  update: function () {
+    return true;
```
Comment (brylie) [inline both/collections/branding.js]: Add check here to make sure user has Role === 'admin'

File: both/collections/branding.js [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+  },
+  remove: function () {
+    return true;
```
Comment (brylie) [inline both/collections/branding.js]: Add check here to make sure user has Role === 'admin'

File: client/layouts/master_layout/master_layout.js [hunk 7]
```diff
@@ -0,0 +1,21 @@
+Template.masterLayout.helpers({
+  branding: function() {
+    var theme = Branding.find();
+    if (theme) {
+      return theme
+    }
+  },
+  projectLogo: function () {
+    var lastUploadedLogo = ProjectLogo.findOne({}, {sort: {uploadedAt: -1}});
```
Comment (brylie) [inline client/layouts/master_layout/master_layout.js]: Add comment describing this line, including the sorting used

File: client/layouts/master_layout/navbar/navbar.html [hunk 8]
```diff
@@ -32,6 +37,7 @@
             <ul class="dropdown-menu">
               <li><a href="{{pathFor 'profile'}}">{{_ "profile"}}</a></li>
               <li><a href="{{pathFor 'account'}}">{{_ "account"}}</a></li>
+              <li><a href="{{pathFor 'branding'}}">Branding</a></li>
```
Comment (brylie) [inline client/layouts/master_layout/navbar/navbar.html]: Make sure user is admin to see this link

File: client/style/global.less [hunk 9]
```diff
@@ -66,6 +66,14 @@ input[name='profile.apiKey']{
   position: inherit !important;
 }
 
+.project-logo {
+  background-repeat: no-repeat;
+  background-position: center;
+  background-size: auto 44px;
+  height: 44px;
+  margin-top: 3px
```
Comment (brylie) [inline client/style/global.less]: Please use relative values such as `em` or `rem` where possible.

File: lib/router/jsonRoutes.js (hunk 3 above)
Comment (brylie) [inline lib/router/jsonRoutes.js]: Make sure we want to merge this file as part of the PR. I.e. we don't want it to break anything relating to the JSON routes.

File: lib/router/main.js [hunk 10]
```diff
@@ -113,4 +113,13 @@ Router.map(function() {
     layoutTemplate: "masterLayout",
     render: "accountsAdmin"
   });
+
+  this.route("branding", {
+    waitOn: function() {
+      return this.subscribe('projectLogo');
```
Comment (brylie) [inline lib/router/main.js]: Manage subscriptions at the template level, rather than the router.

File: client/views/branding/branding.js [hunk 11]
```diff
@@ -0,0 +1,33 @@
+Template.branding.created = function () {
+  // Subscription to branding collection
+  this.subscribe('branding');
```
Comment (brylie) [inline client/views/branding/branding.js]: Also subscribe to logo image here.

```js
let instance = this;

instance.subscribe("projectLogo");
```

**Comment by brylie** [comment]
Really great work @frenchbread. There are a few cleanup tasks, mostly small. Please review.

**Comment by brylie** [comment]
@frenchbread is there a status update for this PR?

**Comment by frenchbread** [comment]
@brylie Added changes. Please review.

File: lib/router/main.js [hunk 12]
```diff
@@ -118,4 +118,14 @@ Router.map(function() {
     layoutTemplate: "masterLayout",
     render: "accountsAdmin"
   });
+
+  this.route("branding", {
+    waitOn: function() {
+      var instance = this;
+      return instance.subscribe('projectLogo');
```
Comment (brylie) [inline lib/router/main.js]: Move this subscription to a template.created() callback, and make sure to remove it from the router waitOn.

File: client/views/branding/branding.js [hunk 13]
```diff
@@ -0,0 +1,35 @@
+Template.branding.created = function () {
+
+  var instance = this;
+  // Subscription to branding collection
+  instance.subscribe('branding');
```
Comment (brylie) [inline client/views/branding/branding.js]: Add `projectLogo` subscription here, and remove it from the router. You can use `instance.subscriptionsReady()` to make sure the subscriptions are loaded before doing certain actions.

**Comment by brylie** [comment]
@frenchbread this PR is very close to being ready to merge. Please move the subscription to the template, and cleanup the router waitOn function.

**Comment by frenchbread** [comment]
Fixes are committed. Please :mag:

File: client/layouts/master_layout/master_layout.js [hunk 14]
```diff
@@ -0,0 +1,24 @@
+Template.masterLayout.helpers({
+  branding: function() {
+    var theme = Branding.find();
```
Comment (brylie) [inline client/layouts/master_layout/master_layout.js]: This is a database cursor. Are you expecting an array of results here?

File: client/views/branding/branding.html [hunk 15]
```diff
@@ -0,0 +1,46 @@
+<template name="branding">
+  <div class="container">
+    <div class="row" id="champion-messages">
+      <div class="col-md-3 settings-menu">
+        <div class="panel panel-default">
+          <div class="panel-body">
+            <ul class="nav nav-pills nav-stacked">
+              <li><a href="#" class="active2">Branding<span class="glyphicon glyphicon-chevron-right pull-right"></span></a></li>
```
Comment (brylie) [inline client/views/branding/branding.html]: Remove the `href` and `class="active2` unless they are used. E.g. is there more than one tab that can be active?

File: client/views/branding/branding.js [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+    var lastUploadedLogo = ProjectLogo.findOne({}, {sort: {uploadedAt: -1}});
+    // Check if new logo was uploaded, if so change it with previous
+    if (lastUploadedLogo) {
+      return lastUploadedLogo
+    }
+  }
+});
+
+
+Template.AdminLTE.helpers({
```
Comment (brylie) [inline client/views/branding/branding.js]: This should be in a different file, since it is not related to the `branding` template. E.g. our convention is to have a template file and javascript file for each template, so there should be:

* adminlte.html
* adminlte.js

**Comment by brylie** [comment]
@frenchbread there are just a few more questions and comments based on my review yesterday. Please let me know if any of these seem unreasonable.

**Comment by brylie** [comment]
@jykae can you help with these couple of cleanup tasks?

**Comment by jykae** [comment]
@brylie Yep, I will assign myself.

File: client/layouts/master_layout/master_layout.js (hunk 14 above)
Comment (jykae) [inline client/layouts/master_layout/master_layout.js]: @brylie @frenchbread I removed the "branding" helper, and didn't break anything. Wasn't used in master_layout.html. Probably something old side path, that was forgotten there :)

File: client/views/branding/branding.html (hunk 15 above)
Comment (jykae) [inline client/views/branding/branding.html]: If remember right there were ideas having "Admin-level" basic settings tab & branding tab in this page. We could leave this for future imho.

"active2" class is has styling that is used and having "href=#" makes the mouse pointer to hand.
I would leave these for now. Let's have a chat tomorrow @brylie :)

File: client/views/branding/branding.js (hunk 16 above)
Comment (jykae) [inline client/views/branding/branding.js]: Moved adminLTE helpers to "api-umbrella-dashboard/client/layouts/master_layout/adminLte.js"

**Comment by jykae** [comment]
@brylie Looks ready for me now. Let's merge this tomorrow.

**Comment by brylie** [comment]
Great work @elnzv, @frenchbread  and @jykae! :+1:
